### PR TITLE
Universal Checkpoint for Sequence Parallelism

### PR DESCRIPTION
**Pull request by samadejacobs**
This PR extends the [universal checkpoint](https://github.com/microsoft/Megatron-DeepSpeed/tree/main/examples_deepspeed/universal_checkpointing) to support DS sequence parallelism and training scenarios where pipeline parallelism is not enabled.

The attached Tensorboard chart show a training scenario (validation curve) where a GPT model is pre-trained with data parallelism (4 GPUs), and checkpoints are saved at the 100th and 200th iterations. The checkpoint at the 100th iteration is later loaded for continual pre-training with different configurations (more GPU resources, data parallelism = 4 GPUs, sequence parallelism = 2 GPUs).

Screenshot 2023-11-28 at 9 11 55 AM

### CODE REVIEW

File: deepspeed/checkpoint/deepspeed_checkpoint.py [hunk 1]
```diff
@@ -32,9 +32,9 @@
 
 class DeepSpeedCheckpoint(object):
 
-    def __init__(self, dir, tp_degree=None, pp_degree=None, dp_degree=None):
+    def __init__(self, dir, tp_degree=None, pp_degree=None, dp_degree=None,pipeline_parallel=False):
```
Comment (tjruwase) [inline deepspeed/checkpoint/deepspeed_checkpoint.py:35]: Can we reuse `pp_degree=None` to represent `pipeline_parallel=False`?

Comment (tjruwase) [reply deepspeed/checkpoint/deepspeed_checkpoint.py:35]: Never mind, that won't work.

File: deepspeed/checkpoint/deepspeed_checkpoint.py [hunk 2]
```diff
@@ -270,12 +276,17 @@ def _merge_state_dicts(self, sd_list):
 
         return merged_sd
 
-    def _validate_folder(self, dir):
+    def _validate_folder(self, dir,pipeline_parallel=False):
         basic_folder_validation(dir)
 
         file_list = get_files(dir)
```
Comment (tjruwase) [inline deepspeed/checkpoint/deepspeed_checkpoint.py:282]: I think we can examine `file_list` to determine if `layer*` files exist and thus distinguish PP and non-PP runs. In general, I don't like the idea of introducing `--pipeline-parallel`.

File: deepspeed/runtime/zero/stage_1_and_2.py [hunk 3]
```diff
@@ -2269,11 +2269,13 @@ def _load_hp_checkpoint_state(self, checkpoint_dir):
...
         tp_rank = bwc_tensor_model_parallel_rank(mpu=self.mpu)
-        tp_world_size = self.mpu.get_slice_parallel_world_size()
+        tp_world_size = self.mpu.get_slice_parallel_world_size() if hasattr(self.mpu, "get_slice_parallel_world_size") \
+                else self.mpu.get_tensor_model_parallel_world_size()
+
         for i, _ in enumerate(self.optimizer.param_groups):
             for lp in self.bit16_groups[i]:
                 if lp._hp_mapping is not None:
-                    #print(f"Loading {self.param_names[lp]} {tp_rank=} {tp_world_size=}")
+                    print(f"Loading {self.param_names[lp]} {tp_rank=} {tp_world_size=}")
```
Comment (tjruwase) [inline deepspeed/runtime/zero/stage_1_and_2.py:2278]: Please keep this commented out.

File: deepspeed/checkpoint/deepspeed_checkpoint.py [hunk 4]
```diff
@@ -32,9 +32,9 @@
 
 class DeepSpeedCheckpoint(object):
 
-    def __init__(self, dir, tp_degree=None, pp_degree=None, dp_degree=None):
+    def __init__(self, dir, tp_degree=None, pp_degree=None, dp_degree=None,pipeline_parallel=False):
         self.dir = dir
-        self._validate_folder(dir)
+        self._validate_folder(dir,pipeline_parallel)
```
Comment (tjruwase) [inline deepspeed/checkpoint/deepspeed_checkpoint.py:38]:
```suggestion
pipeline_parallel = len(get_files_with_prefix(get_files(dir), LAYER_FILE_PREFIX)) > 0
```

File: deepspeed/checkpoint/deepspeed_checkpoint.py [hunk 5]
```diff
@@ -240,8 +252,8 @@ def _build_transformer_file_map(self):
 
     def _sanity_check(self):
         assert len(self.mp_rank_files) % self.tp_degree == 0
-        assert len(self.layer_keys) > 2
-        assert self.zero_checkpoint.num_files % (self.pp_degree * self.tp_degree) == 0
+        #assert self.zero_checkpoint.num_files % (self.pp_degree * self.tp_degree) == 0
```
Comment (tjruwase) [inline deepspeed/checkpoint/deepspeed_checkpoint.py:255]: Should this still be commented out? It seems `self.pp_degree == 1` when there is no pipeline parallelism, right?

File: deepspeed/checkpoint/deepspeed_checkpoint.py [hunk 6]
```diff
@@ -222,9 +232,11 @@ def _build_transformer_file_map(self):
         transformer_layer_keys = self.layer_keys[1:-1]
         file_map = {}
         # XXX: this is not guaranteed
-        layers_per_pp = len(transformer_layer_keys) // self.pp_degree
-        if layers_per_pp == 0:
-            layers_per_pp = 1
+        layers_per_pp = 1
+        if self.pp_degree > 0:
+            layers_per_pp = len(transformer_layer_keys) // self.pp_degree
+        #if layers_per_pp == 0:
```
Comment (tjruwase) [inline deepspeed/checkpoint/deepspeed_checkpoint.py:238]: Remove commented lines.

**Review by tjruwase — APPROVED** [review]

**Review by mrwyattii — APPROVED** [review]

**Pull request merged**